build/toolchain/mac/BUILD.gn

```
@@ -222,7 +222,7 @@ template("mac_toolchain") {
     }
 
     tool("copy") {
-      command = "ln -f {{source}} {{output}} 2>/dev/null || (rm -rf {{output}} && cp -af {{source}} {{output}})"
+      command = "ln -f {{source}} {{output}} 2>/dev/null || (rm -rf {{output}} && rsync -arl {{source}} {{output}})"
```

Instead of `(rm -rf {{output}} && rsync -arl {{source}} {{output}})` 
1. `man rsync` says `-a` is the same as `-rlptgoD` so do we need the `rl`?
2. Instead of the `rm -rf` we can use `rsync --delete` so it will only copy what's needed, but delete anything extraneous (this is what we do in the tool)
https://github.com/flutter/flutter/blob/0a7f8af6d14696e9086cdd390792f7eb8a9cefa5/packages/flutter_tools/lib/src/base/os.dart#L458-L461

```suggestion
      command = "ln -f {{source}} {{output}} 2>/dev/null || (rsync -a --delete {{source}} {{output}})"
```

---

Also, do you know what's up with the symlink part?

---

> so do we need the `rl`?

Good catch; nope, safe to remove.

> do you know what's up with the symlink part?

The code is trying to be clever. The logic is basically:
1. If possible, create a hard link (notice it's not `-s`), deleting any existing file.
2. If that's not possible (e.g. the source file is on a different volume), then `rm -rf` the previous file and copy the file.

This logic definitely predates APFS. With APFS, copies are free (well you still pay the cost of the inodes) and the copy is materialised only if you edit the file. i.e. APFS is effectively copy-on-write.

> Instead of the `rm -rf` we can use `rsync --delete`

Looks like the default behaviour is `--delete-before` when `--delete` is specified, so I _think_ this would preserve the existing behaviour exactly. Will do a quick test locally just out of paranoia.

---

Re: APFS, looking closer at the `cp` code, `cp` only takes advantage of cloned inodes if the [`-c` option is passed](https://github.com/apple-oss-distributions/file_cmds/blob/414172a34456f1b54807c9c0aa5d9200a585333f/cp/utils.c#L289-L297), but falls back to `copyfile()` if `clonefile()` fails, which presumably is under the same conditions as hardlinking would -- i.e. source and destination on a different volume.

I'd have to look to see if `rsync` takes advantage of this or not; there's no mention in the manpage.

---

Based on a really quick grep of `rsync` it doesn't appear Apple is using `clonefile` or passing `COPYFILE_CLONE` etc. to `copyfile` flags so tempted to keep the `ln`.